### PR TITLE
Add ephemeral storage support for JavaScript cookies

### PR DESCRIPTION
**Pull request by mrobinson**
This change adds a new class EphemeralStoragePartition which provides
access to an ephemeral version of a RestrictedCookieManager for the
content process. In an upcoming change this will also provide a
StoragePartition for navigation requests in third-party frames in order
to add support for navigation cookies.

Resolves brave/brave-browser#12591

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`, `npm run gn_check`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Requested a security/privacy review as needed.
- [x] Added appropriate QA labels (`QA/Yes` or `QA/No`) to the associated issue
- [x] Added appropriate release note labels (`release-notes/include` or `release-notes/exclude`) to the associated issue
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: chromium_src/net/cookies/cookie_monster.h
Comment (bridiver) [inline chromium_src/net/cookies/cookie_monster.h:12]: this method doesn't need to be in the header, you can pass in ephemeral_cookie_stores_ to non-class method

File: patches/services-network-public-mojom-cookie_manager.mojom.patch
Comment (bridiver) [inline patches/services-network-public-mojom-cookie_manager.mojom.patch:17]: never add blank lines or comments to patches

File: chromium_src/net/base/url_util.cc
Comment (bridiver) [inline chromium_src/net/base/url_util.cc:92]: I don't believe this is the correct behavior. You should create a URL from the domain even if it's empty and in turn Origin will create an opaque origin for those entries

File: chromium_src/services/network/restricted_cookie_manager.cc
Comment (bridiver) [inline chromium_src/services/network/restricted_cookie_manager.cc:21]: setting the ephemeral storage domain seems to overly complicate things compared to just setting the url and figuring out the storage domain inside the CookieMonster

File: chromium_src/net/base/url_util.cc
Comment (bridiver) [inline chromium_src/net/base/url_util.cc:85]: I don't think this method is needed here https://github.com/brave/brave-core/pull/7154/files#r524857285

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.h
Comment (bridiver) [inline browser/ephemeral_storage/ephemeral_storage_tab_helper.h:47]: why is this a scoped_refptr?

File: patches/services-network-restricted_cookie_manager.cc.patch
Comment (bridiver) [inline patches/services-network-restricted_cookie_manager.cc.patch:9]: this can't work for network cookies so again this doesn't make sense and is not what I described

File: patches/net-cookies-cookie_monster.cc.patch
Comment (bridiver) [inline patches/net-cookies-cookie_monster.cc.patch:5]: I think all of these can be eliminated by subclassing CookieMonster and doing a chromium_src override in content::CreateCookieStore

File: content/browser/tld_ephemeral_storage.cc
Comment (bridiver) [inline content/browser/tld_ephemeral_storage.cc:54]: why does this need to be in content?

File: patches/net-cookies-cookie_monster.cc.patch
Comment (mrobinson) [reply patches/net-cookies-cookie_monster.cc.patch:5]: I did consider subclassing `CookieMonster` and maybe that does make sense. The issue is that `content::CreateCookieStore` is not always used to create the `CookieStore`. Crucially, it isn't used for navigation cookies and instead the constructor is called directly. See `net/url_request/url_request_context_builder.cc` and in `services/network/network_context.cc` (using `make_unique`). If we simply override the factory, there is a chance that this will require more work rebasing onto new versions of Chromium.

File: patches/services-network-restricted_cookie_manager.cc.patch
Comment (mrobinson) [reply patches/services-network-restricted_cookie_manager.cc.patch:9]: Based on your comments on Slack @bridiver, it seems you'd like to have `ComputeSameSiteContextForRequest` do the work of storing the top-frame URL into the CookieOptions.
 1. `ComputeSameSiteContextForRequest` isn't called for cookies set via `document.cookie`. Do you imagine it would be called here?
 2. We will be landing support for network cookies in a future PR.

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.h
Comment (mrobinson) [reply browser/ephemeral_storage/ephemeral_storage_tab_helper.h:47]: Reference-counting is used to manage the lifetime of all ephemeral storage. This matches how the lifetime of storage is matched for DOM storage via `SessionStorageNamespace` in upstream Chromium and in our ephemeral storage implementation.

File: chromium_src/net/base/url_util.cc
Comment (mrobinson) [reply chromium_src/net/base/url_util.cc:85]: Where do you suggest goes in a way that it can be used in the `CookieMonster` and also in `EphemeralTabStorageHelper` or do you prefer to have two version of this code?

File: chromium_src/net/base/url_util.cc
Comment (mrobinson) [reply chromium_src/net/base/url_util.cc:92]: This would mean that all sites that are accessed with IP addresses share the same ephemeral storage. If that the behaviour that we want, I think it makes sense to make that change in a separate PR that also adds tests for that functionality. Since this PR is just about adding support for ephemeral `document.cookie`, I'm not sure it belongs here.

File: chromium_src/net/cookies/cookie_monster.h
Comment (mrobinson) [reply chromium_src/net/cookies/cookie_monster.h:12]: :+1:

File: patches/services-network-public-mojom-cookie_manager.mojom.patch
Comment (mrobinson) [reply patches/services-network-public-mojom-cookie_manager.mojom.patch:17]: :+1:

File: chromium_src/net/cookies/cookie_monster.h
Comment (mrobinson) [reply chromium_src/net/cookies/cookie_monster.h:12]: I don't understand why fewer methods is better here, but I'm happy to make this change.

File: content/browser/tld_ephemeral_storage.cc
Comment (mrobinson) [reply content/browser/tld_ephemeral_storage.cc:54]: This class is the public interface that `EphemeralTabStorageHelper` uses to access private API in content. Is there a better place for this?

File: chromium_src/net/cookies/cookie_monster.h
Comment (mrobinson) [reply chromium_src/net/cookies/cookie_monster.h:12]: This method also uses net_log_ from `CookieMonster` in order to create the ephemeral monster. Since it's using to instance variables, I think it's simpler for it to be a method. Is there a reason why having fewer methods here is better?

File: chromium_src/services/network/restricted_cookie_manager.cc
Comment (mrobinson) [reply chromium_src/services/network/restricted_cookie_manager.cc:21]: :+1:

File: patches/net-cookies-cookie_monster.cc.patch
Comment (mrobinson) [reply patches/net-cookies-cookie_monster.cc.patch:5]: I chatted on slack with @bridiver about this and he clarified that by using chromium_src override, we should be able to create a subclass called `CookieMonster` and then use macros to rename the original CookieMonster. I will give this a shot.

File: patches/services-network-restricted_cookie_manager.cc.patch
Comment (bridiver) [reply patches/services-network-restricted_cookie_manager.cc.patch:9]: as mentioned in slack this was my mistake from searching this page for `BRAVE_SETCANONICALCOOKIE` and not noticing there were two macros with the same name

File: chromium_src/net/base/url_util.cc
Comment (bridiver) [reply chromium_src/net/base/url_util.cc:85]: I don't think we need two versions of the code if we just pass in the url and only convert it to ephemeral storage domain when we actually create it. The code would only ever be called from one file

File: chromium_src/net/base/url_util.cc
Comment (bridiver) [reply chromium_src/net/base/url_util.cc:92]: no, opaque origins are unique by design, but I think actually it would cause the opposite problem where the same IP in more than one tab would go into different ephemeral storage domains. Since url.Getorigin will return an empty url (and hence create an opaque origin) for file urls, I believe you need to call `SerializeWithNonce`

Comment (bridiver) [reply chromium_src/net/base/url_util.cc:92]: what is `domain` for a file url with this code?

File: content/browser/tld_ephemeral_storage.cc
Comment (bridiver) [reply content/browser/tld_ephemeral_storage.cc:54]: what private api?

File: chromium_src/net/base/url_util.cc
Comment (mrobinson) [reply chromium_src/net/base/url_util.cc:85]: This is also called in `EphemeralStorageTabHelper` to manage DOM storage. We need to make sure the conversion from URL to storage domain is in sync.

File: content/browser/tld_ephemeral_storage.cc
Comment (bridiver) [reply content/browser/tld_ephemeral_storage.cc:54]: the only private apis I see are accessed outside of this class and we're now adding two new patches for this. It would be better I think to just keep it all included in the browser_context override. Also brave/content should never have been created in the first place and the files that are already in there are misplaced

Comment (bridiver) [reply content/browser/tld_ephemeral_storage.cc:54]: if you created a separate file for tld_ephemeral_storage.cc/.h and included it from browser_context.cc you would put it in chromium_src

Comment (mrobinson) [reply content/browser/tld_ephemeral_storage.cc:54]: There are three interfaces into private API here CreateSessionStorageNamespace and GetSessionStorageNamespaceId, which used to be in `BrowserContext` (still part of content). The final piece is `TLDEphemeralStorage` which talks to the `CookieStore` in order to manage ephemeral cookie lifetime. The idea is that eventually CreateSessionStorageNamespace and GetSessionStorageNamespaceId would become part of `TLDEphemeralStorage`.

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.h
Comment (bridiver) [reply browser/ephemeral_storage/ephemeral_storage_tab_helper.h:47]: ok, I see what's happening now with the weakptr map

**Comment by mrobinson** [comment]
@bridiver Okay. I've pushed a new commit, which I believe addresses your concerns. PTAL.

File: patches/services-network-public-mojom-cookie_manager.mojom.patch
Comment (bridiver) [inline patches/services-network-public-mojom-cookie_manager.mojom.patch:9]: nit: I don't think we need to specify that this is for ephemeral storage here

File: browser/brave_tab_helpers.cc
Comment (bridiver) [inline browser/brave_tab_helpers.cc:26]: looks like you need to add `//net`

**Comment by mrobinson** [comment]
Okay. I've pushed a new version of the branch which I believe addresses the review comments above. Thanks!

File: chromium_src/content/browser/tld_ephemeral_storage.cc
Comment (iefremov) [inline chromium_src/content/browser/tld_ephemeral_storage.cc:41]: I don't quite follow why do we need a weak pointer instead of a raw one?

Comment (mrobinson) [reply chromium_src/content/browser/tld_ephemeral_storage.cc:41]: In this case the weak pointer is just a sanity check which catches errors where the TLDEphemeralStorage is dereferenced and deleted.

File: chromium_src/content/browser/tld_ephemeral_storage.cc
Comment (iefremov) [inline chromium_src/content/browser/tld_ephemeral_storage.cc:10]: there are many unused headers, probably if we drop them we could move the whole class outside of `chromium_src`?

File: chromium_src/net/cookies/cookie_monster.h
Comment (iefremov) [inline chromium_src/net/cookies/cookie_monster.h:9]: `ChromiumCookieMonster`?

File: chromium_src/net/cookies/cookie_monster.cc
Comment (iefremov) [inline chromium_src/net/cookies/cookie_monster.cc:8]: not needed?

File: chromium_src/net/cookies/cookie_monster.cc
Comment (iefremov) [inline chromium_src/net/cookies/cookie_monster.cc:47]: no need for this check, `emplace` will do the trick on its own

File: chromium_src/content/public/browser/tld_ephemeral_storage.h
Comment (iefremov) [inline chromium_src/content/public/browser/tld_ephemeral_storage.h:35]: I'm not sure about future plans for this class, but for now it doesn't look like a "storage", rather something like `TLDEphemeralLifetime`

File: chromium_src/content/browser/tld_ephemeral_storage.cc
Comment (iefremov) [reply chromium_src/content/browser/tld_ephemeral_storage.cc:41]: which checks do you have in mind?

**Comment by iefremov** [comment]
Assuming that the patching work is negotiated with @bridiver and most probably cannot be done any better, the whole thing looks good. I'll try to give it test, but for now it failed to build on top of master

**Comment by iefremov** [comment]
(I also assume that commit messages will be updated before merging)

File: chromium_src/content/browser/tld_ephemeral_storage.cc
Comment (mrobinson) [reply chromium_src/content/browser/tld_ephemeral_storage.cc:41]: `.get()` on the weak pointer will return nullptr instead of a garbage value, but I will also add a DCHECK to also produce an assertion failure.

File: chromium_src/content/browser/tld_ephemeral_storage.cc
Comment (mrobinson) [reply chromium_src/content/browser/tld_ephemeral_storage.cc:10]: I will definitely remove the unused headers, but the idea is that eventually the private API exposed in browser_context.h can move into this class and it can eventually handle keeping track of all TLD-associated ephemeral DOM storage as well.

File: chromium_src/net/cookies/cookie_monster.cc
Comment (mrobinson) [reply chromium_src/net/cookies/cookie_monster.cc:47]: Won't this result in creating another instantiation of `ChromiumCookieMonster` every time we call this method?

**Comment by mrobinson** [comment]
> Assuming that the patching work is negotiated with @bridiver and most probably cannot be done any better, the whole thing looks good. I'll try to give it test, but for now it fails to build on top of master

I've pushed a rebased version of this patch and as well a commit addressing review comments.

File: chromium_src/content/public/browser/tld_ephemeral_storage.h
Comment (mrobinson) [reply chromium_src/content/public/browser/tld_ephemeral_storage.h:35]: I haven't changed this yet, because perhaps it doesn't make sense to rename this class and then rename it back in a followup change. I'm happy to keep discussing it though.

File: chromium_src/net/cookies/cookie_monster.cc
Comment (iefremov) [reply chromium_src/net/cookies/cookie_monster.cc:46]: yeah, it seems we should leave it alone

File: chromium_src/content/browser/tld_ephemeral_storage.cc
Comment (iefremov) [reply chromium_src/content/browser/tld_ephemeral_storage.cc:33]: using `WeakPtr` in this way sounds unusual to me, but probably fine - could you write a comment describing this?

File: chromium_src/content/public/browser/tld_ephemeral_storage.h
Comment (iefremov) [reply chromium_src/content/public/browser/tld_ephemeral_storage.h:35]: But even with encapsulated DOMStorage - would it still be "storage"? For now we basically only keep namespaces and manage their lifetime in the tab helper, instead of "storing" something

Comment (mrobinson) [reply chromium_src/content/public/browser/tld_ephemeral_storage.h:35]: Okay. I will change the name.

File: chromium_src/content/browser/tld_ephemeral_storage.cc
Comment (mrobinson) [reply chromium_src/content/browser/tld_ephemeral_storage.cc:33]: Okay. I'll leave a comment.

**Comment by mrobinson** [comment]
I've pushed a commit which I believe addresses the remaining review comments.

**Comment by mrobinson** [comment]
It seems that the Windows build has an issue with patches to mojom trait headers. I've taken a different tactic, which is to extend `CookieStore` API to add two special methods for ephemeral cookie access. This also has the added benefit of reducing the number of Chromium patches necessary by 2. PTAL.

**Review by iefremov — APPROVED** [review]
pls mention a link to the wiki explainer somewhere in the code (e.g. ephemeral_storage_tab_helper.h and/or ephemeral_storage_lifetime.h)

File: chromium_src/content/public/browser/browser_context.h
Comment (bridiver) [inline chromium_src/content/public/browser/browser_context.h:9]: what is this change? Doesn't seem related and also should have 4 space indent for continuation of a line

File: chromium_src/content/browser/tld_ephemeral_lifetime.cc
Comment (bridiver) [inline chromium_src/content/browser/tld_ephemeral_lifetime.cc:1]: this doesn't belong in chromium_src because it's not overriding a chromium file

File: chromium_src/content/browser/tld_ephemeral_lifetime.cc
Comment (bridiver) [inline chromium_src/content/browser/tld_ephemeral_lifetime.cc:6]: do not use brave/chromium_src in the include path

File: chromium_src/content/browser/browser_context.cc
Comment (bridiver) [inline chromium_src/content/browser/browser_context.cc:48]: do not use brave/chromium_src in the include path

File: chromium_src/content/public/browser/browser_context.h
Comment (bridiver) [inline chromium_src/content/public/browser/browser_context.h:15]: don't use brave/chromium_src in the include path

File: chromium_src/content/public/browser/browser_context.h
Comment (mrobinson) [reply chromium_src/content/public/browser/browser_context.h:9]: There have been changes to the lint script which mean that the CI fails unless you do a full format. This will possibly be resolved when https://github.com/brave/brave-core/pull/7402 lands.

Comment (mrobinson) [reply chromium_src/content/public/browser/browser_context.h:9]: In the meantime, I can remove these full format changes.

**Review by bridiver — APPROVED** [review]
approved pending the few small fixes

**Comment by mrobinson** [comment]
There are three failures in CI:
 * Linux and Android: This is failing in the lint stage (while successful in build and testing). Lint has passed on other platforms and locally. There seem to be issues with the new lint script running on CI and the advice given to me is that this can be ignored if it's unrelated to my change.
 * iOS: It seems that the `AdsWrapperTest testPreferencePersistance` test is failing. This is a known issue and the current advice is to ignore it. Finally, this change should be unrelated to that test failure.

Given this, I feel fairly confident about merging this change despite the CI failures.

**Pull request merged**
